File: Data Structures & Algorithms/permutation-string/submission-1.rs
impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let s1 = s1.as_bytes();
        let s2 = s2.as_bytes();
        let window_size = s1.len();

        if window_size > s2.len() {
            return false;
        }

        let mut required = [0i32; 26];
        let mut window = [0i32; 26];

        for i in 0..window_size {
            required[(s1[i] - b'a') as usize] += 1;
            window[(s2[i] - b'a') as usize] += 1;
        }

        if window == required {
            return true;
        }

        for right in window_size..s2.len() {
            window[(s2[right] - b'a') as usize] += 1;
            window[(s2[right - window_size] - b'a') as usize] -= 1;

            if window == required {
                return true;
            }
        }

        false
    }
}

File: Data Structures & Algorithms/minimum-window-with-characters/submission-1.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not t or len(t) > len(s):
            return ""

        required = [0] * 128
        window = [0] * 128
        needed_types = 0

        for character in t:
            index = ord(character)
            if required[index] == 0:
                needed_types += 1
            required[index] += 1

        formed = 0
        left = 0
        best_start = 0
        best_length = len(s) + 1

        for right, character in enumerate(s):
            index = ord(character)
            window[index] += 1

            if required[index] > 0 and window[index] == required[index]:
                formed += 1

            while formed == needed_types:
                length = right - left + 1
                if length < best_length:
                    best_start = left
                    best_length = length

                left_index = ord(s[left])
                if required[left_index] > 0 and window[left_index] == required[left_index]:
                    formed -= 1
                window[left_index] -= 1
                left += 1

        if best_length == len(s) + 1:
            return ""

        return s[best_start:best_start + best_length]